plotly/graph_reference.py

```diff
@@ -0,0 +1,64 @@
+"""
+This module handles accessing, storing, and managing the graph reference.
+
+"""
+from __future__ import absolute_import
+
+import hashlib
+import json
+import os
+import requests
+from pkg_resources import resource_string
+
+import six
+
+from plotly import exceptions, files, utils
+
+GRAPH_REFERENCE_PATH = '/v2/plot-schema'
+
+
+def get_graph_reference():
+    """
+    Attempts to load local copy of graph reference or makes GET request if DNE.
+
+    :return: (dict) The graph reference.
+    :raises: (PlotlyError) When graph reference DNE and GET request fails.
+
+    """
+    default_config = files.FILE_CONTENT[files.CONFIG_FILE]
+    if files.check_file_permissions():
+        graph_reference = utils.load_json_dict(files.GRAPH_REFERENCE_FILE)
+        config = utils.load_json_dict(files.CONFIG_FILE)
+
+        # TODO: https://github.com/plotly/python-api/issues/293
+        plotly_api_domain = config.get('plotly_api_domain',
+                                       default_config['plotly_api_domain'])
+    else:
+        graph_reference = {}
+        plotly_api_domain = default_config['plotly_api_domain']
+
+    sha1 = hashlib.sha1(six.b(str(graph_reference))).hexdigest()
+
+    graph_reference_url = '{}{}?sha1={}'.format(plotly_api_domain,
+                                                GRAPH_REFERENCE_PATH, sha1)
```

Comment:
Presumably, when plotly offline downloads (the latest) `plotly.js`, it can also download that version's `plot-schema`. This schema won't need updating, since it is locked in with the local `plotly.js`.